src/IO/H5/VolumeData.cpp

```diff
@@ -154,6 +162,308 @@ void append_element_extents_and_connectivity(
+  for (size_t i = 0; i < block_number_for_each_element.size(); ++i) {
+    sorted_element_indices[block_number_for_each_element[i]].push_back(i);
+  }
+
+  return std::make_tuple(number_of_blocks, block_number_for_each_element,
+                         sorted_element_indices);
+}
+
+// Sort input by block
+template <typename T>
+std::vector<std::vector<T>> sort_by_block(
+    const std::vector<std::vector<size_t>>& sorted_element_indices,
+    const std::vector<T>& property_to_sort) {
+  std::vector<std::vector<T>> sorted_property;
+  sorted_property.reserve(sorted_element_indices.size());
+
+  // Properly sizes subvectors
+  for (const auto& sorted_block_index : sorted_element_indices) {
+    std::vector<T> sizing_vector;
+    sizing_vector.reserve(sorted_block_index.size());
+    for (const auto& sorted_element_index : sorted_block_index) {
+      sizing_vector.push_back(property_to_sort[sorted_element_index]);
+    }
+    sorted_property.push_back(sizing_vector);
+  }
+
+  return sorted_property;
+}
+
+// Returns properties for each block
+template <size_t SpatialDim>
+std::tuple<size_t, size_t, std::array<int, SpatialDim>>
+compute_block_level_properties(
+    const std::vector<std::string>& block_grid_names,
+    const std::vector<std::vector<size_t>>& block_extents) {
+  size_t expected_connectivity_length = 0;
+  size_t expected_number_of_grid_points = 0;
+
+  for (const auto& extents : block_extents) {
+    size_t element_grid_points = 1;
+    size_t element_connectivity_length = 1;
+    for (size_t j = 0; j < SpatialDim; j++) {
+      element_grid_points *= extents[j];
+      element_connectivity_length *= extents[j] - 1;
+    }
+    // Connectivity that already exists
+    expected_connectivity_length += element_connectivity_length * 8;
+    // Used for reserving the length of block_logical_coords
+    expected_number_of_grid_points += element_grid_points;
+  }
+
+  std::string grid_name_string = block_grid_names[0];
+  std::array<int, SpatialDim> h_ref_array = {};
+  size_t h_ref_previous_start_position = 0;
+  size_t additional_connectivity_length = 1;
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    size_t h_ref_start_position =
+        grid_name_string.find('L', h_ref_previous_start_position + 1);
+    size_t h_ref_end_position =
+        grid_name_string.find('I', h_ref_start_position);
+    int h_ref = std::stoi(
+        grid_name_string.substr(h_ref_start_position + 1,
+                                h_ref_end_position - h_ref_start_position - 1));
+    h_ref_array[i] = h_ref;
+    additional_connectivity_length *= pow(2, h_ref + 1) - 1;
+    h_ref_previous_start_position = h_ref_start_position;
+  }
+
+  expected_connectivity_length +=
+      (additional_connectivity_length - block_extents.size()) * 8;
+
+  return std::tuple{expected_connectivity_length,
+                    expected_number_of_grid_points, h_ref_array};
+}
+
+// Generates the mesh
+template <size_t SpatialDim>
+Mesh<SpatialDim> generate_element_mesh(
```

Comment:
Do you think it would be useful to add this into `NumericalAlgorithms/Spectra/Mesh.?pp` with the name like `mesh_from_strings()`? I could see this function coming in handy for other people

Reply:
Yea! If you think it could be useful for others, I could move it to a separate preliminary commit (or into the same commit as the GenerateXdmf change) and just call the function from that other file. Is there a format (commit wise) that you would prefer?

Reply:
Yea, I think doing it as a separate commit with its own test is the right way to go :) Also means you can submit that as a separate PR if we end up finding that's best :)

Reply:
Sounds good! Should the plan be to get this PR fixed up and eventually merged as is, and then a follow-up PR focused on disentangling this function? Or vice versa?

Reply:
Would it make sense to factor out this functionality and do a PR for it now, making the final PR for ExtendConnectivity lighter? Or is it fine to just factor it in this PR and move it into its own commit?

Reply:
Just a note for me to remember: I think this function is now already in develop